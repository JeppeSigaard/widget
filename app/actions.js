import { BASEURL, APPID } from "../config";
import axios from "axios";
import { stringify } from "query-string";

// Action constant
export const SET_WEATHER = "SET_WEATHER";

// Set weather
export const setWeather = function(status, data) {
  return { type: SET_WEATHER, status, data };
};

// Get weather
export const getWeather = function(city) {
  return (dispatch, getState) => {
    const query = stringify({ q: `${city},dk`, appid: APPID, units: "metric" });
    dispatch(setWeather("loading"));

    axios
      .get(`${BASEURL}?${query}`)
      .then(response => {
        dispatch(setWeather("success", response.data));
      })
      .catch(e => {
        dispatch(setWeather("failed", e));
      });
  };
};
