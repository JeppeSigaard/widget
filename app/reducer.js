export default function(state = {}, action) {
  if (action.type === "SET_WEATHER") {
    switch (action.status) {
      case "loading": {
        return state;
      }

      case "failed": {
        return { ...state, cod: "404" };
      }

      case "success": {
        return action.data;
      }
    }
  }
  return state;
}
