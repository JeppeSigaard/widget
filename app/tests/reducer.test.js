import reducer from "../reducer";
import { SET_WEATHER } from "../actions";

const mockActionLoading = {
  type: SET_WEATHER,
  status: "loading",
  data: { sum: "otherdata" }
};

const mockActionFailed = {
  type: SET_WEATHER,
  status: "failed",
  data: { sum: "data" }
};

const mockAction = {
  type: SET_WEATHER,
  status: "success",
  data: { sum: "data" }
};

describe("Reducer", () => {
  it("has initial state", () => {
    expect(reducer(undefined, {})).toEqual({});
  });

  it("does not update state on loading", () => {
    expect(reducer({ sum: "data" }, mockActionLoading)).toMatchSnapshot();
  });

  it("sets status code on error", () => {
    expect(reducer({}, mockActionFailed)).toMatchSnapshot();
  });

  it("stores weather data", () => {
    expect(reducer({}, mockAction)).toMatchSnapshot();
  });
});
