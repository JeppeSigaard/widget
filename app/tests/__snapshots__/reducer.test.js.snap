// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer does not update state on loading 1`] = `
Object {
  "sum": "data",
}
`;

exports[`Reducer sets status code on error 1`] = `
Object {
  "cod": "404",
}
`;

exports[`Reducer stores weather data 1`] = `
Object {
  "sum": "data",
}
`;
